Support consistent hashing in soft affinity node selection strategy

rongrong opened the pull request:
```
== RELEASE NOTES ==

General Changes
* Add support to use consistent hashing as a hash strategy when affinity scheduling is used. This feature can be turned on by setting ``node-scheduler.node-selection-hash-strategy`` to ``CONSISTENT``. A separate configuration ``node-scheduler.consistent-hashing-min-virtual-node-count`` specifies the minimal number of virtual nodes in the consistent hashing ring.
```

rongrong commented:
Based on earlier work in #14738

rongrong commented:
cc @maobaolong

beinan commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java:
```diff
@@ -277,7 +297,7 @@ public NodeSelector createNodeSelector(Session session, ConnectorId connectorId,
             }
 
-            return new NodeMap(activeNodesByNodeId.build(), activeWorkersByNetworkPath.build(), coordinatorNodeIds, new LinkedList<>(activeNodes), new LinkedList<>(allNodes), allNodesByHost.build(), allNodesByHostAndPort.build());
+            return new NodeMap(activeNodesByNodeId.build(), activeWorkersByNetworkPath.build(), coordinatorNodeIds, new LinkedList<>(activeNodes), new LinkedList<>(allNodes), allNodesByHost.build(), allNodesByHostAndPort.build(), activeNodesByConsistentHashing);
```
split the line?

beinan commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java:
```diff
@@ -147,24 +152,21 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
                     break;
                 case SOFT_AFFINITY:
-                    candidateNodes = selectExactNodes(nodeMap, split.getPreferredNodes(allCandidates), includeCoordinator);
+                    if (nodeSelectionHashStrategy.equals(MODULAR)) {
```
nit: using == for enum,  concise and null safe

beinan commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSchedulerConfig.java:
```diff
@@ -37,6 +37,8 @@
     private int maxPendingSplitsPerTask = 10;
     private int maxUnacknowledgedSplitsPerTask = 500;
     private String networkTopology = NetworkTopologyType.LEGACY;
+    private NodeSelectionHashStrategy nodeSelectionHashStrategy = NodeSelectionHashStrategy.MODULAR;
```
Is there any reason for us to keep the modular strategy in the future?

rongrong replied:
We can delete modular strategy later once this is tested and rolled out fully. But to be able to safely test it we need both. And keeping the original behavior as default would make sure the feature is not accidentally on during deployment.

yuanzhanhku replied:
Agreed with Rongrong. Rolling out the new strategy will invalidate most of the existing local caches due to the shuffle of the nodes. We would like to roll it out in a more controlled manner and active monitor the clusters after it is released.

yuanzhanhku commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java:
```diff
@@ -257,6 +270,13 @@ public NodeSelector createNodeSelector(Session session, ConnectorId connectorId,
                     .map(InternalNode::getNodeIdentifier)
                     .collect(toImmutableSet());
 
+            int weight = (int) ceil(1.0 * minVirtualNodeCount / activeNodes.size());
```
Where is the logic of compute the actual virtual node count based on minVirtualNodeCount  and physical node count?

rongrong replied:
"weight" is basically the number of virtual nodes created for each physical node. So the actual number of virtual node is `weight * physicalNodeCount = ceil(minVirtualNodeCount / physicalNodeCount) * physicalNodeCount >= minVirtualNodeCount`. I figured it's better to make the number of virtual node configurable, but then the exact number is probably not important, mostly just the order of magnitudes. Directly configuring the weight is another option, but then the desired weight could be very different for a 10 node cluster vs a 500 node cluster.

kewang1024 replied:
Why don't we directly change the config to `node-scheduler.consistent-hashing-virtual-node-count` to represent the weight here? 
so it would be an independent config from how many nodes a cluster have

rongrong replied:
Because, as I mentioned above, if you set weight directly, the actual number of virtual nodes on a hashing ring will depend on the number of physical node, which means it's hard to give a good default weight setting.

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java:
```diff
@@ -147,24 +146,16 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
                     break;
                 case SOFT_AFFINITY:
-                    candidateNodes = selectExactNodes(nodeMap, split.getPreferredNodes(allCandidates), includeCoordinator);
+                    candidateNodes = selectExactNodes(nodeMap, split.getPreferredNodes(new ModularHashingNodeProvider(nodeMap.getAllNodes())), includeCoordinator);
```
We lost the sorting logic in this change

rongrong replied:
Where is the sorting logic?

kewang1024 replied:
we removed Line 150 ~ 157 and it has the sorting logic

rongrong replied:
I don't see sorting logic there.

rongrong replied:
Seems like the sorting logic was removed in https://github.com/prestodb/presto/pull/15791

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java:
```diff
@@ -257,6 +270,13 @@ public NodeSelector createNodeSelector(Session session, ConnectorId connectorId,
                     .map(InternalNode::getNodeIdentifier)
                     .collect(toImmutableSet());
 
+            int weight = (int) ceil(1.0 * minVirtualNodeCount / activeNodes.size());
+            for (InternalNode node : activeNodes) {
+                for (int i = 0; i < weight; i++) {
+                    activeNodesByConsistentHashing.put(murmur3_32().hashString(format("%s%d", node.getNodeIdentifier(), i), UTF_8).asInt(), node);
+                }
+            }
```
We should move those logic inside `ConsistentHashingNodeProvider` given they're specific to `ConsistentHashingNodeProvider` and NodeScheduler shouldn't be aware of it

rongrong replied:
In that case the map will be created for every query. Here it's created only once when node map is updated.

kewang1024 replied:
That's a good point, then should we actually only have one  `NodeProvider ` in the system instead of creating one for each query

rongrong replied:
I thought about that. Basically we only need to create a `NodeProvider` when `NodeMap` is changed. Currently this is cumbersome to do because when we use soft affinity with modular hashing, we are using all nodes, rather than just active nodes, to reduce reshuffling due to temporary worker unavailability.  Once re remove modular hashing, and always use active nodes as candidate nodes, the `NodeProvider` logic can be moved into `NodeMap` entirely.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/NodeProvider.java:
```diff
@@ -0,0 +1,26 @@
+     * @param count        how many desirable nodes to be returned
+     * @return a list of the chosen nodes by specific hash function
+     */
+    List<HostAddress> get(String identifier, int count);
```
Generally speaking, I prefer to only add capabilities when they are used. So when we have another use case for consistent hashing that requires generics, we can always refactor and change that. Or if you have more details of the routing logic, I'd love to learn more.

kewang1024 replied:
Come to think of it, we also need to restructure the routing logic accordingly to adapt the consistent hashing, so yeah we can defer this change when we want to revamp affinity routing

rongrong commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeMap.java:
```diff
@@ -84,4 +91,16 @@ public NodeMap(
+        switch (nodeSelectionHashStrategy) {
+            case MODULAR:
+                return new ModularHashingNodeProvider(activeNodes);
+            case CONSISTENT:
```
The class is called `HashStrategy` already. If you think it could be confusing, then to make it uniform, should we name them `MODULAR_HASHING` and `CONSISTENT_HASHING`?

kewang1024 replied:
> The class is called `HashStrategy` already. If you think it could be confusing, then to make it uniform, should we name them `MODULAR_HASHING` and `CONSISTENT_HASHING`?

yeah, sounds good, that would be better

kewang1024 commented on presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java:
```diff
@@ -577,6 +585,57 @@ public void testAffinityAssignment()
     }
 
+    @Test
+    public void testAffinityAssignmentWithConsistentHashing()
```
Can we also add test for the virtual node? right now in this test, it would also generate one place in the hashing ring for each node

rongrong replied:
It's a bit harder to reason when more nodes. I'll try.

kewang1024 reviewed (commented):
Can we add test for the `ConsistentHashingNodeProvider` to test the even distribution?

We have an test example in gateway, shared in here: https://docs.google.com/document/d/1C_kMut0XS8T3pBYkbAE-xsef0_206XCmMpPzbEs9Q8w/edit?usp=sharing

rongrong commented:
> Can we add test for the `ConsistentHashingNodeProvider` to test the even distribution?
> 
> We have an test example in gateway, shared in here: https://docs.google.com/document/d/1C_kMut0XS8T3pBYkbAE-xsef0_206XCmMpPzbEs9Q8w/edit?usp=sharing

Is the test stable enough? I'm hesitant on introducing non-deterministic tests in unit test. Distribution is an attribute of the hashing function we choose. So it's not really testing our logic. Is it really necessary?

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/ModularHashingNodeProvider.java:
```diff
@@ -0,0 +1,52 @@
+        int position = mod < 0 ? mod + size : mod;
+        List<HostAddress> chosenCandidates = new ArrayList<>();
+        for (int i = 0; i < count && i < sortedCandidates.size(); i++) {
+            chosenCandidates.add(sortedCandidates.get((position + i) % size).getHostAndPort());
```
(position + i) % size could be negative, we need to add extra logic to handle that

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java:
```diff
@@ -146,16 +152,21 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
                     break;
                 case SOFT_AFFINITY:
-                    candidateNodes = selectExactNodes(nodeMap, split.getPreferredNodes(new ModularHashingNodeProvider(nodeMap.getAllNodes())), includeCoordinator);
+                    if (nodeSelectionHashStrategy == MODULAR_HASHING) {
+                        nodeProvider = new ModularHashingNodeProvider(nodeMap.getAllNodes());
+                    }
```
I believe this part is the hacky override to make sure we're compatible with old implementation?  Can we add a comment or todo here just in case people would get confused

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeMap.java:
```diff
@@ -84,4 +91,16 @@ public NodeMap(
         return allNodesByHostAndPort;
     }
+
+    public NodeProvider getActiveNodeProvider(NodeSelectionHashStrategy nodeSelectionHashStrategy)
+    {
+        switch (nodeSelectionHashStrategy) {
+            case MODULAR_HASHING:
+                return new ModularHashingNodeProvider(activeNodes);
+            case CONSISTENT_HASHING:
+                return new ConsistentHashingNodeProvider(activeNodesByConsistentHashing);
+            default:
+                throw new IllegalArgumentException(format("Unknown NodeSelectionHashStrategy: %s", nodeSelectionHashStrategy));
+        }
+    }
```
Why do we create one for each query? we can move the creation into `NodeMap` constructor, and `getActiveNodeProvider` only return the corresponding instance

In this way, we can move the specific node logic into `ConsistentHashingNodeProvider` (the one we were discussing in NodeScheduler.java)

rongrong replied:
Sure, but this constructor is really cheap. It doesn't really matter.

kewang1024 commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeMap.java:
```diff
@@ -84,4 +91,16 @@ public NodeMap(
+    {
+        switch (nodeSelectionHashStrategy) {
+            case MODULAR_HASHING:
+                return new ModularHashingNodeProvider(activeNodes);
```
Why don't we just use `allNodes` to keep the original logic? 
then we can also remove the hacky part 166~168 in SimpleNodeSelector.java

rongrong replied:
The original logic uses `activeNodes` for hard affinity, and `allNodes` only with soft affinity. `ActiveNodes` should be the right one to use. `allNodes` is the hack.

rongrong commented:
I added a test to test hash distribution. It's not very uniform. The node identifier and split identifier used in the test are not realistic. I don't know "random" the actual values would be and whether those would generate a better distribution.

rongrong commented:
Had a chat with @kewang1024  offline today. I addressed the comments. There are still 2 open questions and 1 follow up:
Question 1: should we configure # virtual node in the hashing ring, or the weight of each physical node, or both.
Question 2: Choice of hashing function and distribution of hash values. I added a test to test hash distribution with fake node identifiers and fake split path. The hashing values of these fake values are not very nicely uniformly distributed. I’m not sure whether actual values would fair better on this.
Follow up: once consistent hashing is working and rolled out, we should probably delete modular hashing. There’s a hack on modular hashing to use all nodes for hashing rather than just consider active nodes, which makes the implementation of this quite awkward. If we don’t have good reasons to keep modular hashing, maybe we should just remove it later, thus removing the hack.

beinan reviewed (approved):
lgtm, thanks!
